app/services/product_tag_rules_filterer.rb

```diff
@@ -0,0 +1,111 @@
+# Takes a Spree::Variant AR object and filters results based on applicable tag rules.
+# Tag rules exists in the context of enterprise, customer, and variant_overrides,
+# and are applied to variant_overrides only. Returns a Spree::Variant AR object.
+
+class ProductTagRulesFilterer
+  def initialize(distributor, customer, variants_relation)
+    @distributor = distributor
+    @customer = customer
+    @variants_relation = variants_relation
+  end
+
+  def call
+    return variants_relation unless distributor_rules.any?
+
+    filter!(variants_relation)
```

Comment:
I would remove the bang as it doesn't follow any of the meanings it could have in Ruby land: either a method that performs a mutation (Ruby) or a method that may raise an exception (Rails). IMO it can be very confusing for anyone used to Ruby.

Reply:
Interestingly all the services that deal with tag rules seem have a `#filter!` method with a bang. I've removed it here. :+1: